### PR TITLE
added a 4x TS Demo

### PR DESCRIPTION
**Pull request by odoe**
Simple TypeScript demo app using 4x. Just highlights how to use typings and includes `npm` scripts to compile while watching for file changes.

Please review @dasa

### CODE REVIEW

File: 4.x/typescript/demo/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Demo App for using TypeScript with [ArcGIS API 4 for JavaScript](https://developers.arcgis.com/javascript/)
+
+## Usage
+
+Clone the repo and run `npm install`
+
+To start working run `npm run dev`
```
Comment (dasa) [inline 4.x/typescript/demo/README.md]: Maybe add a `lint` script and mention it here?

File: 4.x/typescript/demo/package.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "demo",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "dev": "tsc -w -p ."
```
Comment (dasa) [inline 4.x/typescript/demo/package.json]: I don't think you need `-p .`

File: 4.x/typescript/demo/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    "noImplicitAny": true,
+    "outDir": "app",
+    "removeComments": false,
+    "sourceMap": true,
+    "jsx": "react",
+    "reactNamespace": "jsxFactory",
+    "target": "es5",
+    "isolatedModules": false,
+    "experimentalDecorators": true,
+    "emitDecoratorMetadata": true,
```
Comment (dasa) [inline 4.x/typescript/demo/tsconfig.json]: Is this needed?

File: 4.x/typescript/demo/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    "declaration": false,
+    "noLib": false,
+    "preserveConstEnums": true,
+    "suppressImplicitAnyIndexErrors": true
+  },
```
Comment (dasa) [inline 4.x/typescript/demo/tsconfig.json]: I'd remove the properties that are set to `false` unless you think someone may want to set them to `true`.

File: 4.x/typescript/demo/tsconfig.json (hunk 3 above)
Comment (odoe) [inline 4.x/typescript/demo/tsconfig.json]: I thought it was a good start if someone uses it as a base for widgets. I can remove it though.

File: 4.x/typescript/demo/tsconfig.json (hunk 4 above)
Comment (odoe) [inline 4.x/typescript/demo/tsconfig.json]: cool

File: 4.x/typescript/demo/package.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "demo",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
Comment (dasa) [inline 4.x/typescript/demo/package.json]: Remove `main`

Comment (odoe) [inline 4.x/typescript/demo/package.json]: nice catch!

File: 4.x/typescript/demo/package.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  },
+  "keywords": ["esri", "typescript"],
+  "author": "Rene Rubalcava <rrubalcava@esri.com>",
+  "license": "Apache 2.0",
+  "devDependencies": {
+    "tslint": "^4.2.0",
+    "typescript": "^2.1.4"
+  },
+  "dependencies": {
+    "@types/arcgis-js-api": "^4.2.0"
```
Comment (dasa) [inline 4.x/typescript/demo/package.json]: Consolidate to `dependencies`

File: 4.x/typescript/demo/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,20 @@
+{
+  "compilerOptions": {
+    "module": "amd",
+    "noImplicitAny": true,
+    "outDir": "app",
```
Comment (dasa) [inline 4.x/typescript/demo/tsconfig.json]: `outDir` seems redundant since ts files are inside `app`.

**Pull request merged**